### PR TITLE
Update basic04

### PR DESCRIPTION
**Pull request by matsduf**
## Purpose

The purpose of this PR is to add a specific message when IPv4 or IPv6 has been disabled. The default message is too verbose and to be kept on DEBUG level.

Editorial updates have also been done. Below it is specified what changes have any impact on the implementation in Zonemaster::Engine.

* Updated specification - b729a2d96408 - *This update has no impact on implementation.*
  * Updates to match template.
  * No logical changes.
  * No changes that requires updated implementation.
  * msgid has been copied from implementation.
* Editorial update: adjusted link list - 07089023013d7 - *This update has no impact on implementation.*
* Adds reference to "DNS Query and Response Defaults" but no change of logic. - a001b793184b80 - *This update has no impact on implementation.*
* Adds specific messages when IPv4 or IPv6 has been disabled. - b5d62a2a4 - **This update requires an update to the implementation.**

## Context

zonemaster/zonemaster-engine/pull/790 increased the verbosity of the standard messages, which makes it necessary to move them to the DEBUG level also for Basic.

## How to test this PR

This is a document change, and the resulting update in implementation should be tested.

### CODE REVIEW

File: docs/specifications/tests/Basic-TP/basic04.md [hunk 1]
```diff
@@ -23,46 +37,97 @@ report problems found in the following areas:
 * Name Server responding with unexpected RCODE (any except "NOERROR")
```
Comment (tgreenx) [inline docs/specifications/tests/Basic-TP/basic04.md:37]: RCODE -> [RCODE Name] (with link to IANA RCODE Name page)
"NOERROR" -> "NoError"

Comment (tgreenx) [reply docs/specifications/tests/Basic-TP/basic04.md:37]: This also applies in the Test Procedure below where I can see two other occurrences

File: docs/specifications/tests/Basic-TP/basic04.md [hunk 2]
```diff
@@ -3,6 +3,20 @@
...
+* [Test procedure](#Test-procedure)
+* [Outcome(s)](#Outcomes)
+* [Special procedural requirements](#Special-procedural-requirements)
+* [Intercase dependencies](#Intercase-dependencies)
+* [Terminology](#terminology)
+
+
 ## Objective
 
 Many test cases will query the name servers in the delegation or the
```
Comment (tgreenx) [inline docs/specifications/tests/Basic-TP/basic04.md:22]: SOA record or both. Reporting problem is crucial -> SOA record, or both. Reporting problems is crucial

**Review by tgreenx — COMMENTED** [review]
LGTM

File: docs/specifications/tests/Basic-TP/basic04.md (hunk 1 above)
Comment (matsduf) [reply docs/specifications/tests/Basic-TP/basic04.md:37]: Fixed three places plus added link in link section.

File: docs/specifications/tests/Basic-TP/basic04.md (hunk 2 above)
Comment (matsduf) [reply docs/specifications/tests/Basic-TP/basic04.md:22]: Fixed.

File: docs/specifications/tests/Basic-TP/basic04.md [hunk 3]
```diff
@@ -34,7 +34,7 @@ report problems found in the following areas:
   section in the response on an NS query for *Child Zone*.
 * Name Server not setting the AA flag in a response with SOA or NS in
   answer section.
-* Name Server responding with unexpected RCODE (any except "NOERROR")
+* Name Server responding with unexpected [RCODE Name] (any except "NeError")
```
Comment (tgreenx) [inline docs/specifications/tests/Basic-TP/basic04.md:37]: NeError -> NoError

File: docs/specifications/tests/Basic-TP/basic04.md [hunk 4]
```diff
@@ -21,7 +21,7 @@
 
 Many test cases will query the name servers in the delegation or the
 name servers appointed by the NS records in the zone for the NS or SOA
-record or both. Reporting problem is crucial, but instead of letting
+record, or both. Reporting problem is crucial, but instead of letting
```
Comment (tgreenx) [inline docs/specifications/tests/Basic-TP/basic04.md:24]: Reporting problem -> Reporting problems

Comment (matsduf) [reply docs/specifications/tests/Basic-TP/basic04.md:24]: I do not like "problems is" so I rephrased it.

File: docs/specifications/tests/Basic-TP/basic04.md (hunk 3 above)
Comment (matsduf) [reply docs/specifications/tests/Basic-TP/basic04.md:37]: Fixed.

**Review by tgreenx — APPROVED** [review]

File: docs/specifications/tests/Basic-TP/basic04.md [hunk 5]
```diff
@@ -20,105 +34,166 @@ report problems found in the following areas:
...
-   2. If there is no DNS response on neither query (UDP), then:
-      1. Send the SOA query over TCP to the name server and collect the
-         response.
-      2. If there is no DNS response on the TCP query, then output
-         *[B04_NO_RESPONSE]* and go to next server.
-      3. Else (there is a DNS response over TCP), then output 
-         *[B04_RESPONSE_TCP_NOT_UDP]* and go to next server.
+4. If IPv4 is disabled then do:
+   1. Extract all name servers with IPv4 address from *Name Server IP*.
+   2. If the set of IPv4 name serververs is non-empty then output
```
Comment (pnax) [inline docs/specifications/tests/Basic-TP/basic04.md:100]: typo: "serververs" -> "servers"

File: docs/specifications/tests/Basic-TP/basic04.md [hunk 6]
```diff
@@ -20,105 +34,166 @@ report problems found in the following areas:
...
+      *[B04_IPV4_DISABLED]* with the set of IPv4 name servers.
+
+5. If IPv6 is disabled then do:
+   1. Extract all name servers with IPv6 address from *Name Server IP*.
+   2. If the set of IPv6 name serververs is non-empty then output
```
Comment (pnax) [inline docs/specifications/tests/Basic-TP/basic04.md:105]: typo: "serververs" -> "servers"

File: docs/specifications/tests/Basic-TP/basic04.md [hunk 7]
```diff
@@ -20,105 +34,166 @@ report problems found in the following areas:
...
+      *[B04_IPV6_DISABLED]* with the set of IPv6 name servers.
+
+6. For each name server in *Name Server IP* do:
+
+   1. Send *SOA Query UDP* and *NS Query UDP* to the name server and collect
+      the [DNS Responses][DNS Response].
+   2. If there is no DNS response on neither query), then:
```
Comment (pnax) [inline docs/specifications/tests/Basic-TP/basic04.md:112]: typo: closing parenthesis

File: docs/specifications/tests/Basic-TP/basic04.md [hunk 8]
```diff
@@ -20,105 +34,166 @@ report problems found in the following areas:
...
+      1. Send *SOA Query TCP* to the name server and collect the [DNS Response].
+      2. If there is no [DNS Response], then output *[B04_NO_RESPONSE]* and go
+         to next server.
+      3. Else, then output *[B04_RESPONSE_TCP_NOT_UDP]* and go to next server.
```
Comment (pnax) [inline docs/specifications/tests/Basic-TP/basic04.md:116]: Why going to next server and not checking the response over TCP in the rest of the procedure? It feels like the procedure is missing the case where the nameserver responds only over TCP with unexpected responses.

File: docs/specifications/tests/Basic-TP/basic04.md (hunk 5 above)
Comment (matsduf) [reply docs/specifications/tests/Basic-TP/basic04.md:100]: What a typo! Fixed.

File: docs/specifications/tests/Basic-TP/basic04.md (hunk 6 above)
Comment (matsduf) [reply docs/specifications/tests/Basic-TP/basic04.md:105]: Well, I guess copy and paste of the previous one. Fixed.

File: docs/specifications/tests/Basic-TP/basic04.md (hunk 7 above)
Comment (matsduf) [reply docs/specifications/tests/Basic-TP/basic04.md:112]: Removed.

File: docs/specifications/tests/Basic-TP/basic04.md (hunk 8 above)
Comment (matsduf) [reply docs/specifications/tests/Basic-TP/basic04.md:116]: A name server must respond over UDP, and DNS resolvers will generally send queries over UDP only, and only use TCP of the response was truncated (TC flag set). In the following test cases of Zonemaster all queries will be sent of UDP (one exceptional test case testing queries of TCP). I do not feel that adding tests of TCP will add anything substantial here. The server is unreachable, in a practical sense. The DNS operator must fix that.

After reading my response I felt that the message ID for B04_RESPONSE_TCP_NOT_UDP should probably be more explicit about the issue. Please review my update.

Comment (pnax) [reply docs/specifications/tests/Basic-TP/basic04.md:116]: I understand however I still think we are missing an edge case (that could be easily addressed I think). But I won't bother if we agree this is fine like that.

Comment (matsduf) [reply docs/specifications/tests/Basic-TP/basic04.md:116]: We could let the others express their opinion too. It is of course very possible to do what you suggest.

Comment (matsduf) [reply docs/specifications/tests/Basic-TP/basic04.md:116]: I think it would make more sense to update [Connectivity02] ("TCP connectivity") to only test those name servers that do respond over UDP, but then test them thoroughly over TCP, both SOA and NS query, give correct rcode and have AA flag set.

[Connectivity02]: https://github.com/zonemaster/zonemaster/blob/develop/docs/specifications/tests/Connectivity-TP/connectivity02.md

Comment (tgreenx) [reply docs/specifications/tests/Basic-TP/basic04.md:116]: I feel like, since this test case is the base test case in testing a name server functionality, specifications (RFCs) should prevail rather than operational usage. I couldn't find a quote from an RFC (i.e [RFC1035 ](https://www.rfc-editor.org/rfc/rfc1035.html), [RFC7766](https://www.rfc-editor.org/rfc/rfc7766.html)) that clearly indicated that a name server should (or must) respond to UDP if it already responds on TCP. But your point is also valid @matsduf ("DNS resolvers will generally send queries over UDP only"). Thus I think we can proceed with @pnax suggestion and use that TCP response for further testing when UDP is unavailable, while still outputting a comprehensive message tag B04_RESPONSE_TCP_NOT_UDP to warn the user of possible reachability shortcomings.

Comment (matsduf) [reply docs/specifications/tests/Basic-TP/basic04.md:116]: Supporting both UDP and TCP is mandatory. "Section 6.1.3.2 of RFC1123 is updated: All general-purpose DNS implementations MUST support both UDP and TCP transport." (https://www.rfc-editor.org/rfc/rfc7766#section-5)

Comment (tgreenx) [reply docs/specifications/tests/Basic-TP/basic04.md:116]: Right, thanks, somehow I missed that...
From the same section ([Section 5](https://www.rfc-editor.org/rfc/rfc7766#section-5)), it follows with:
```
   Regarding the choice of when to use UDP or TCP, Section 6.1.3.2 of
   RFC 1123 also says:

      ... a DNS resolver or server that is sending a non-zone-transfer
      query MUST send a UDP query first.

   This requirement is hereby relaxed.  Stub resolvers and recursive
   resolvers MAY elect to send either TCP or UDP queries depending on
   local operational reasons.  TCP MAY be used before sending any UDP
   queries.  If the resolver already has an open TCP connection to the
   server, it SHOULD reuse this connection.  In essence, TCP ought to be
   considered a valid alternative transport to UDP, not purely a retry
   option.
```

My understanding is that a DNS response over TCP, independently from UDP, must still be considered valid. So it ought to count for something when it is found. Thus I stand with my previous opinion (i.e., use that TCP response for further testing and warn the user of major issue with lack of UDP support + likelihood of reachability issues that it entails)

Comment (matsduf) [reply docs/specifications/tests/Basic-TP/basic04.md:116]: A DNS response over TCP is valid, if the client sent the query over TCP. The client can choose to send all queries over TCP, but the server must accept both. A DNS resolver must accept queries over both from its clients, but it can choose to send its queries, as client, over TCP.

The background to Basic04 is that many test cases outputted NO_RESPONSE (or similar) to the same name server when failing to connect. It is or was not very helpful to get the same message multiple times. In all these case the message was outputted because connection over **UDP** was failing. I then created Basic04 to give one such message, and since it replaced multiple messages over failing connection over UDP it is natural that Basic04 has a UDP focus. All testcases, with two exceptions (if I remember correctly), only queries over UDP.

I added the extra TCP test when connection over UDP failed, but it can be questioned if that is really relevant for the purpose of the test case. I added that to be helpful in case the name server had UDP blocked by a firewall.

I suggest that we remove the TCP test from Basic04 and then update update  [Connectivity02] ("TCP connectivity") and let it test all name servers:

1. If the name server has no UDP connection only look for TCP as Basic04 does today.
2. If the name server has UDP connection do a thorough test over TCP, both SOA and NS query, give correct RCODE and have AA flag set.

[Connectivity02]: https://github.com/zonemaster/zonemaster/blob/develop/docs/specifications/tests/Connectivity-TP/connectivity02.md

**Comment by matsduf** [comment]
Replaced by #1097, #1098 and #1099.
